index.bs

```diff
@@ -101,21 +118,190 @@ spec: webidl
+      <a>current realm</a> permission to access. Each of these features defines
+      an <a>extra permission data type</a>, and then a {{PermissionName}} |name|'s
+      <dfn export>extra permission data</dfn> is the instance of that type
+      matching the UA's impression of the user's intent. Successive uses of
+      |name|'s <a>extra permission data</a> should return the same value, unless
+      the UA receives new information about the user's intent.
+    </p>
+  </section>
+  <section>
+    <h3 id="requesting-more-permission">Requesting more permission</h3>
+
+    <div algorithm="request-permission-to-use">
+      <p>
+        To <dfn export>request permission to use</dfn> a |descriptor|, the UA
+        must perform the following steps. Note that these steps may wait for
+        user input, so they should not be used from other algorithms running on
+        the main thread. This algorithm returns either {{"granted"}} or
+        {{"denied"}}.
+      </p>
+      <ol>
+        <li>
+          If |descriptor|'s <a>permission state</a> is not {{"prompt"}}, return
+          that value and abort these steps.
+        </li>
+        <li>
+          Show the user a prompt asking their permission for the calling
+          algorithm to use powerful feature described by |descriptor|.
+        </li>
+        <li>
+          If the user grants permission, return {{"granted"}}; otherwise return
+          {{"denied"}}. Depending on the details of the user's interaction, the
+          UA may also treat this as new information about the user's intent for
+          other realms with the same origin.
+
+          <p class="note">
+            This is intentionally vague about the details of the permission UI
+            and how the UA infers user intent. UAs should be able to explore
+            lots of UI within this framework.
+          </p>
+        </li>
+      </ol>
+    </div>
+
+    <div algorithm="prompt-user-to-choose">
+      <p>
+        To <dfn export>prompt the user to choose</dfn> one of several options
+        associated with a |descriptor|, the UA must perform the following steps.
+        Note that these steps may wait for user input, so they should not be
+        used from other algorithms running on the main thread. This algorithm
+        returns either {{"denied"}} or one of the options.
+      </p>
+      <ol>
+        <li>
+          If |descriptor|'s <a>permission state</a> is {{"denied"}},
+          return {{"denied"}} and abort these steps.
+        </li>
+        <li>
+          If |descriptor|'s <a>permission state</a> is {{"granted"}}, the UA may
+          return one of the options and abort these steps. If the UA returns
+          without prompting, then successive <a lt="prompt the user to
+          choose">prompts for the user to choose</a> from the same set of
+          options with the same |descriptor| should return the same option,
```

Comment:
Changed to 'must'. I don't think constraining changes to turn boundaries is particularly useful to authors, since this algorithm's running in parallel anyway and so you can't arrange to have two calls to it happen within the same turn. I tried to guarantee this in #96 if you want to take a look.